// @mojotech/json-type-validation/dist/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Result = require("./result");
exports.Result = Result;
var decoder_1 = require("./decoder");
exports.Decoder = decoder_1.Decoder;
exports.isDecoderError = decoder_1.isDecoderError;
var combinators_1 = require("./combinators");
exports.string = combinators_1.string;
exports.number = combinators_1.number;
exports.boolean = combinators_1.boolean;
exports.anyJson = combinators_1.anyJson;
exports.unknownJson = combinators_1.unknownJson;
exports.constant = combinators_1.constant;
exports.object = combinators_1.object;
exports.array = combinators_1.array;
exports.tuple = combinators_1.tuple;
exports.dict = combinators_1.dict;
exports.optional = combinators_1.optional;
exports.oneOf = combinators_1.oneOf;
exports.union = combinators_1.union;
exports.intersection = combinators_1.intersection;
exports.withDefault = combinators_1.withDefault;
exports.valueAt = combinators_1.valueAt;
exports.succeed = combinators_1.succeed;
exports.fail = combinators_1.fail;
exports.lazy = combinators_1.lazy;
